File: backend/api/tables.py
"""DXF 表格抽取相关 API：单表 + 批量异步 + 查询。"""
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from backend.core.database import get_db
from backend.schemas.background_job import BackgroundJobStatus
from backend.schemas.drawing_table import (
    DrawingTableRecord,
    ExtractTablesBatchRequest,
    ExtractTablesRequest,
    ExtractTablesSummary,
)
from backend.services import cad_table_service

router = APIRouter(prefix="/api", tags=["tables"])


@router.post("/sheets/{sheet_id}/extract-tables", response_model=ExtractTablesSummary)
def extract_tables(
    sheet_id: int,
    payload: ExtractTablesRequest | None = None,
    db: Session = Depends(get_db),
) -> ExtractTablesSummary:
    return cad_table_service.extract_tables_from_sheet(db, sheet_id, payload=payload)


@router.post("/imports/{batch_id}/extract-tables", response_model=BackgroundJobStatus)
def extract_tables_batch(
    batch_id: int,
    payload: ExtractTablesBatchRequest | None = None,
    db: Session = Depends(get_db),
) -> BackgroundJobStatus:
    return cad_table_service.start_extract_tables_batch_job(
        db, batch_id, payload or ExtractTablesBatchRequest()
    )


@router.get(
    "/imports/{batch_id}/extract-tables/job",
    response_model=BackgroundJobStatus | None,
)
def get_extract_tables_job(
    batch_id: int,
    db: Session = Depends(get_db),
) -> BackgroundJobStatus | None:
    return cad_table_service.get_extract_tables_job(db, batch_id)


@router.get("/sheets/{sheet_id}/tables", response_model=list[DrawingTableRecord])
def list_sheet_tables(
    sheet_id: int,
    db: Session = Depends(get_db),
) -> list[DrawingTableRecord]:
    return cad_table_service.list_tables_for_sheet(db, sheet_id)


@router.get("/projects/{project_id}/tables", response_model=list[DrawingTableRecord])
def list_project_tables(
    project_id: int,
    kind: str | None = Query(default=None, description="按表格类型筛选"),
    db: Session = Depends(get_db),
) -> list[DrawingTableRecord]:
    return cad_table_service.list_tables_for_project(db, project_id, kind=kind)

File: frontend/src/api/tables.ts
import { apiGet, apiPost } from "./client";

export type TableExtractionMethod = "acad_table" | "text_cluster";
export type TableKind =
  | "equipment"
  | "material"
  | "drawing_index"
  | "legend"
  | "other";

export type DrawingTableRecord = {
  id: number;
  project_id: number;
  batch_id: number;
  file_id: number;
  sheet_id: number;
  table_index: number;
  extraction_method: TableExtractionMethod;
  table_kind: TableKind;
  layer_name: string | null;
  header: string[];
  rows: string[][];
  row_count: number;
  col_count: number;
  source_bbox: number[] | null;
  warnings: string[];
  created_at: string;
  updated_at: string;
};

export type ExtractTablesRequest = {
  include_acad_table?: boolean;
  include_text_cluster?: boolean;
  force?: boolean;
};

export type ExtractTablesBatchRequest = ExtractTablesRequest & {
  continue_on_error?: boolean;
};

export type ExtractTablesSummary = {
  sheet_id: number;
  acad_table_count: number;
  text_cluster_count: number;
  total_count: number;
  by_kind: Record<string, number>;
};

export type TableJobResultSummary = {
  batch_id: number;
  total_sheets: number;
  success_count: number;
  failed_count: number;
  by_kind: Record<string, number>;
};

export type TableBackgroundJobStatus = {
  id: number;
  job_type: string;
  scope_type: "batch" | "project" | "sheet";
  scope_id: number;
  status: "running" | "completed" | "failed";
  total: number;
  processed: number;
  current_step: string | null;
  message: string | null;
  started_at: string;
  finished_at: string | null;
  result_summary: TableJobResultSummary | null;
};

export function extractTablesForSheet(
  sheetId: number,
  payload?: ExtractTablesRequest
): Promise<ExtractTablesSummary> {
  return apiPost<ExtractTablesSummary>(
    `/api/sheets/${sheetId}/extract-tables`,
    payload ?? {}
  );
}

export function startExtractTablesBatch(
  batchId: number,
  payload?: ExtractTablesBatchRequest
): Promise<TableBackgroundJobStatus> {
  return apiPost<TableBackgroundJobStatus>(
    `/api/imports/${batchId}/extract-tables`,
    payload ?? {}
  );
}

export function getExtractTablesJob(
  batchId: number
): Promise<TableBackgroundJobStatus | null> {
  return apiGet<TableBackgroundJobStatus | null>(
    `/api/imports/${batchId}/extract-tables/job`
  );
}

export function listSheetTables(sheetId: number): Promise<DrawingTableRecord[]> {
  return apiGet<DrawingTableRecord[]>(`/api/sheets/${sheetId}/tables`);
}

export function listProjectTables(
  projectId: number,
  kind?: TableKind
): Promise<DrawingTableRecord[]> {
  const query = kind ? `?kind=${encodeURIComponent(kind)}` : "";
  return apiGet<DrawingTableRecord[]>(`/api/projects/${projectId}/tables${query}`);
}
